libafl/src/executors/sand.rs

```diff
@@ -0,0 +1,236 @@
+        _state: &mut S,
+        _mgr: &mut EM,
+        _input: &I,
+    ) -> Result<ExitKind, Error> {
+        Ok(ExitKind::Ok)
+    }
+}
+
+impl<Head, Tail, EM, I, S, Z> ExecutorsTuple<EM, I, S, Z> for (Head, Tail)
+where
+    Head: Executor<EM, I, S, Z>,
+    Tail: ExecutorsTuple<EM, I, S, Z>,
+{
+    fn run_targets(
+        &mut self,
+        fuzzer: &mut Z,
+        state: &mut S,
+        mgr: &mut EM,
+        input: &I,
+    ) -> Result<ExitKind, Error> {
+        let kind = self.0.run_target(fuzzer, state, mgr, input)?;
+        if kind == ExitKind::Crash {
+            return Ok(kind);
+        }
+        self.1.run_targets(fuzzer, state, mgr, input)
+    }
+}
+
+/// The execution pattern of the SANDExecutor. The default value used in our paper is
+/// [SANDExecutionPattern::SimplifiedTrace] and we by design don't include coverage
+/// increasing pattern here as it will miss at least 25% bugs and easy enough to implement
+/// by iterating the crash corpus.
+#[derive(Debug, Clone, Default, Copy)]
+pub enum SANDExecutionPattern {
+    /// The simplified trace, captures ~92% bug triggering inputs with less than ~5% overhead on average.
+    #[default]
+    SimplifiedTrace,
+    /// The unique trace, captures ~99.9% bug-triggering inputs with more than ~50% overhead.
+    UniqueTrace,
+}
+
+/// The core SANDExecutor. It wraps another executor and a list of extra executors.
+/// Please refer to [SAND.md](https://github.com/AFLplusplus/AFLplusplus/blob/stable/docs/SAND.md) for
+/// how to build `sand_executors`.
+#[derive(Debug, Clone)]
+pub struct SANDExecutor<E, ET, C, O> {
+    executor: E,
+    sand_executors: ET,
+    bitmap: Vec<u8>,
+    ob_ref: Handle<C>,
+    pattern: SANDExecutionPattern,
+    ph: PhantomData<O>,
+}
+
+impl<E, ET, C, O> SANDExecutor<E, ET, C, O>
+where
+    C: Named,
+{
+    fn bitmap_set(&mut self, idx: usize) {
+        let bidx = idx % 8;
+        let idx = (idx / 8) % self.bitmap.len();
+        *self.bitmap.get_mut(idx).unwrap() |= 1u8 << bidx;
+    }
+
+    fn bitmap_read(&mut self, idx: usize) -> u8 {
+        let bidx = idx % 8;
+        let idx = (idx / 8) % self.bitmap.len();
+        (self.bitmap[idx] >> bidx) & 1
+    }
+
+    /// Create a new [SANDExecutor]
+    pub fn new(
+        executor: E,
+        sand_extra_executors: ET,
+        observer_handle: Handle<C>,
+        bitmap_size: usize,
+        pattern: SANDExecutionPattern,
+    ) -> Self {
+        Self {
+            executor,
+            sand_executors: sand_extra_executors,
+            bitmap: vec![0; bitmap_size],
+            ob_ref: observer_handle,
+            pattern,
+            ph: PhantomData::default(),
+        }
+    }
+
+    /// Create a new [SANDExecutor] using paper setup
+    pub fn new_paper(executor: E, sand_extra_executors: ET, observer_handle: Handle<C>) -> Self {
+        Self::new(
+            executor,
+            sand_extra_executors,
+            observer_handle,
+            1 << 29,
```

Comment:
what does 1<<29 mean?

Reply:
In our canonical implementation, the hash is 32 bits and thus `1<<29` `u8` will have `1<<32` bits so that the hash value itself won't have any collision.